feature(notifications): refactor notification system for improved usability

hypeJunction opened the pull request:
Improves usability of instant and subscription notifications:
- Adds "origin" to notification object parameters making it easier to distinguish between
  instant and subscription notifications
- Adds a high level "prepare, notification" plugin hook that can be used to filter
  both subscription and instant notifications.
  Normalizes parameters sent to the hook for both subscription types.
- Adds "format, notification:$method" hook, which can be used to format notification subject
  and/or body, to make it suitable for the delivery method. For example, plugins may strip
  HTML tags from notifcation body for plaintext emails, or wrap the body in an HTML template
  for HTML emails.
- Instant notifications are now sent by the notifications service triggering
  a set of hooks similar to subscriptions notifications.
- Help developers debug notifications by logging notifications and notification
  event results
- [ ] Tests will have to be rewritten after #10076

hypeJunction committed:
feat(notifications): adds elgg_get_notification_methods()

Adds API to retrieve registered notification methods

hypeJunction committed:
feature(notifications): refactor notification system for improved
usability

Improves usability of instant and subscription notifications:

- Adds "origin" to notification object parameters making it easier to distinguish between
instant and subscription notifications

- Adds a high level "prepare, notification" plugin hook that can be used to filter
both subscription and instant notifications.
Normalizes parameters sent to the hook for both subscription types.

- Adds "format, notification:$method" hook, which can be used to format notification subject
and/or body, to make it suitable for the delivery method. For example, plugins may strip
HTML tags from notifcation body for plaintext emails, or wrap the body in an HTML template
for HTML emails.

- Instant notifications are now sent by the notifications service triggering
a set of hooks similar to subscriptions notifications.

- Help developers debug notifications by logging notifications and notification
event results

mrclay commented on docs/guides/upgrading.rst:
```diff
@@ -9,6 +9,15 @@ See the administrator guides for :doc:`how to upgrade a live site </admin/upgrad
+-------------
+
+ * A high level ``'prepare','notification'`` hook is now triggered for instant and subscription notifications and can be used to alter notification objects irrespective of their type.
+ * ``'format','notification:<method>'`` hook is now triggered for instant and subscription notifications and can be used to format the notification (e.g. strip HTML tags, wrap the notification body in a template etc).
```
This hook isn't listed under the Instant Notifications heading

mrclay commented on docs/guides/hooks-list.rst:
```diff
@@ -304,6 +293,122 @@ Permission hooks
 
 
+Subscriptions Notifications
+===========================
```
We should say these are listed chronologically in the lifetime of the notification event, and similarly under the other heading.

mrclay commented on engine/lib/notification.php:
```diff
@@ -504,7 +505,24 @@ function notify_user($to, $from, $subject, $message, array $params = array(), $m
+						$params['sender'] = $sender;
+						$params['recipient'] = $recipient;
+
 						$notification = new \Elgg\Notifications\Notification($sender, $recipient, $language, $subject, $message, $summary, $params);
```
The constructor $params argument receives a bunch of new keys here. Is that intentional? If not I guess I'd prefer to not pass extraneous stuff into APIs. Like `$hook_params = $params` and use that for the new keys.

hypeJunction replied:
Intentional. I want to match them to subscriptions notifications.

mrclay commented on docs/guides/hooks-list.rst:
```diff
@@ -304,6 +293,122 @@ Permission hooks
+    ``$params`` array includes:
+
+     * ``notification`` - a notification object ``\Elgg\Notifications\Notification``
+
```
Might be worth mentioning `[email, system]` here since it's very common.

hypeJunction replied:
I thought about moving the `email,system` hook description down here.

juho-jaakkola commented:
Instead of this I would like to make instant notifications work more like async notifications, so plugin devs wouldn't need to know the origin in the first place.

At some point Cash had the idea of running also instant notifications through the same notifications system as async ones. This way they could make use of all the existing hooks already available in the async notifications system.

hypeJunction commented:
1. Will you have time to work on it before 2.3? I don't want to wait another year to do basic operations on notifications.
2. Either way, I think triggering high level hooks makes sense. You can't possibly know which events, type/subtype combinations trigger events. Using `'prepare', 'all'` causes collisions with the menu preparation hook. So we need a way to work with notifications broadly.

juho-jaakkola commented:
> Will you have time to work on it before 2.3? I don't want to wait another year to do basic operations on notifications.

What are the use cases? Would the hooks triggered by `\Elgg\Notifications\NotificationsService` be enough?

 - `'prepare', 'notification:<action>:<entity_type>:<entity_subtype>'`
 - `'send', 'notification:<method>'`

juho-jaakkola commented:
Might make sense to list the needs and use cases to their own issue (if one doesn't yet exist).

hypeJunction commented:
1. There is no `prepare` hook triggered for instant notifications.
2. Granular hooks can not be triggered for all instant notifications.
3. Send hook is used to actually deliver a notification, so it's not suitable for altering the notification body.
4. If we add generic hooks for the above, it would only make sense they also apply for subscription notifications.

hypeJunction commented:
@cash When you have a minute, could you share your thoughts on the instant notification architecture?

juho-jaakkola commented:
Here is the issue made by Cash: https://github.com/Elgg/Elgg/issues/5691

hypeJunction commented:
Here is what I have come up with.

hypeJunction commented on docs/guides/upgrading.rst:
```diff
@@ -31,6 +31,12 @@ Plugins should use the following convenience functions to build responses:
+
+ * A high level ``'prepare','notification'`` hook is now triggered for instant and subscription notifications and can be used to alter notification objects irrespective of their type.
+ * ``'format','notification:<method>'`` hook is now triggered for instant and subscription notifications and can be used to format the notification (e.g. strip HTML tags, wrap the notification body in a template etc).
+
```
Document new instant notifications hooks

hypeJunction commented on engine/classes/Elgg/Notifications/Event.php:
```diff
@@ -1,12 +1,14 @@
  * @subpackage Notifications
  */
-class Event {
+class Event implements NotificationEvent {
```
@mrclay Any way to make this into a `SubscriptionNotificationEvent` without breaking BC?

mrclay replied:
If the goal is to keep this `Event` class for BC, just make a `SubscriptionNotificationEvent` that simply extends `Event` and use it instead.

hypeJunction commented on engine/classes/Elgg/Notifications/Event.php:
```diff
@@ -32,7 +34,14 @@ class Event {
 	 */
-	public function __construct(\ElggData $object, $action, \ElggEntity $actor = null) {
+	public function __construct(\ElggData $object = null, $action = null, \ElggEntity $actor = null) {
+		if (!isset($object)) {
```
Check `ElggData` instance instead

hypeJunction commented on engine/classes/Elgg/Notifications/NotificationsService.php:
```diff
@@ -49,20 +59,19 @@ class NotificationsService {
+			Queue $queue, PluginHooksService $hooks,
+			ElggSession $session,
+			Translator $translator,
+			EntityTable $entities) {
```
Inject `Logger` and log sent notifications with level `NOTICE` or `INFO`?

hypeJunction commented on engine/classes/Elgg/Notifications/NotificationsService.php:
```diff
@@ -85,7 +94,7 @@ public function registerEvent($type, $subtype, array $actions = array()) {
 		}
 
-		$action_list =& $this->events[$type][$subtype];
+		$action_list = & $this->events[$type][$subtype];
```
MEH

hypeJunction commented on engine/classes/Elgg/Notifications/NotificationsService.php:
```diff
@@ -147,19 +156,17 @@ public function getMethods() {
-			if (isset($this->events[$object_type])
-				&& isset($this->events[$object_type][$object_subtype])
-				&& in_array($action, $this->events[$object_type][$object_subtype])) {
+			if (isset($this->events[$object_type]) && isset($this->events[$object_type][$object_subtype]) && in_array($action, $this->events[$object_type][$object_subtype])) {
```
Shorten this with `!empty`

hypeJunction commented:
@Elgg/contributors If anyone has objections to this direction please speak up now. I will be adding thorough tests for it next - don't want the effort to be wasted.

hypeJunction commented on docs/guides/hooks-list.rst:
```diff
@@ -311,6 +290,131 @@ Permission hooks
+
+	``$params`` may vary based on the notification type and may include:
+
+	 * ``event`` - ``\Elgg\Notifications\NotificationEvent`` instance that describes the notification event. Can be ``null`` for instant notifications.
```
No longer ``null``.

hypeJunction commented on docs/guides/hooks-list.rst:
```diff
@@ -311,6 +290,131 @@ Permission hooks
+
+	``$params`` include:
+
+	 * ``event`` - ``\Elgg\Notifications\NotificationEvent`` instance that describes the notification event. Can be ``null`` for instant notifications.
```
No longer ``null``

hypeJunction commented on docs/guides/notifications.rst:
```diff
@@ -87,7 +87,8 @@ This is the workflow of the notifications system:
  #. Plugins are allowed to alter the subscriptions using the ``[get, subscriptions]`` hook
- #. Plugins are allowed to alter the notification parameters with the ``[send:before, notifications]`` hook
+ #. Plugins are allowed to terminate notifications queue processing with the ``[send:before, notifications]`` hook
+ #. Plugins are allowed to alter the notification parameters with the ``[prepare, notification]`` hook
```
Missing `format` hook

mrclay commented on engine/classes/Elgg/Notifications/Event.php:
```diff
@@ -110,6 +119,23 @@ public function getAction() {
+		$obj = new \stdClass();
+		$vars = get_object_vars($this);
+		foreach ($vars as $key => $value) {
+			if (is_callable([$value, 'toObject'])) {
```
Make sure $value is an object first. If the value happens to match a classname this could cause trouble.

mrclay commented on engine/lib/notification.php:
```diff
@@ -441,81 +441,10 @@ function _elgg_notify_user($to, $from, $subject, $message, array $params = null,
  * @throws NotificationException
  */
 function notify_user($to, $from, $subject, $message, array $params = array(), $methods_override = "") {
-
```
I like all this red here.

mrclay commented on engine/classes/Elgg/Notifications/NotificationsService.php:
```diff
@@ -227,99 +241,262 @@ public function processQueue($stopTime) {
 			$notification = $this->getDeprecatedNotificationBody($notification, $event, $method);
 		}
 
+		$notification = _elgg_services()->hooks->trigger('format', "notification:$method", [], $notification);
```
`$this->hooks`

mrclay commented on engine/classes/Elgg/Notifications/NotificationsService.php:
```diff
@@ -227,99 +241,262 @@ public function processQueue($stopTime) {
+
+		$notification = new Notification($actor, $recipient, $language, $subject, $body, $summary, $params);
+
+		$notification = _elgg_services()->hooks->trigger('prepare', 'notification', $params, $notification);
```
`$this->hooks`

mrclay commented on engine/classes/Elgg/Notifications/InstantNotificationEvent.php:
```diff
@@ -0,0 +1,122 @@
+		$obj = new \stdClass();
+		$vars = get_object_vars($this);
+		foreach ($vars as $key => $value) {
+			if (is_callable([$value, 'toObject'])) {
```
make sure `is_object($value)`

mrclay commented on engine/classes/Elgg/Notifications/Notification.php:
```diff
@@ -93,6 +97,30 @@ public function getRecipient() {
+		unset($vars['action']);
+		unset($vars['object']);
+		foreach ($vars as $key => $value) {
+			if (is_callable([$value, 'toObject'])) {
```
make sure `is_object($value)`

juho-jaakkola commented:
The direction of this PR is looking great!

jdalsem commented:
Keep going with this... much needed

hypeJunction committed:
chore(notifications): extend the scope of subscription notification tests

Extends the scope of subscription notification tests. Fixes some bugs and annoyances found
along the way

hypeJunction commented:
Subscription notifications are now covered by tests for all object types.
Will work on instant notification tests next.

hypeJunction commented:
@mrclay I am curious as to why we rely on unserialized event data after dequeueing the event from the database. Wouldn't it be logical to validate that the object/actor still exist? We blindly inject them back into the notification service.

hypeJunction commented on engine/classes/Elgg/Notifications/Event.php:
```diff
@@ -67,30 +70,16 @@ public function getActor() {
+	 * @return ElggData
 	 */
 	public function getObject() {
-		switch ($this->object_type) {
```
I guess that's the part where we were validating the existence. We can't test this unless we use EntityMocks strategy for relationships and extenders.

mrclay commented:
> why we rely on unserialized event data after dequeueing the event from the database.

@cash may know. I suspect this was meant to hold only simple structures, not entities. Maybe it was done to avoid overhead.

> Wouldn't it be logical to validate that the object/actor still exist?

Yes, though we can address later I think.

hypeJunction commented:
Object and actor were re-instantiated in the `Event`, which I thought I was odd. I will instead add some logic to `NotificationService::enqueueEvent` to validate parameters before instantiating `Event`

hypeJunction committed:
chore(tests): tests can now mock entity metadata, annotations and
relationships

Extends \Elgg\Tests\EntityMocks to allow tests to mock entity metadata,
annotations and relationships

hypeJunction committed:
chore(tests): break down notifications service tests by event object type

hypeJunction committed:
chore(notifications): more reliable subsription event serialization

Notification event serialization is now more reliable and tested.

hypeJunction commented:
`\Elgg\Tests\EntityMocks` can now be used to mock entity metadata, annotations and relationships, and cover basic get/set/delete operations in the database tables.

Each notification object type now has their own test case running for all notifications service tests.

Enqueue/dequeue tests now test how events are serialized/unserialized, so we have more consistency with how they would be stored in the database.

hypeJunction committed:
chore(tests): adds instant notification tests

hypeJunction commented:
I am done here in case anyone wants to get their hands dirty.

hypeJunction commented on engine/lib/notification.php:
```diff
@@ -680,9 +619,15 @@ function elgg_send_email($from, $to, $subject, $body, array $params = null) {
 	$method = get_input('method');
 
-	$current_settings = get_user_notification_settings();
+	$current_settings = $user->getNotificationSettings();
```
cast to object

mrclay commented:
Epic.

hypeJunction commented:
I can make some time to finish this next week, just need eyes and feedback. @elgg/contributors

hypeJunction committed:
chore(notifications): correctly retrieve array value

ElggUser::getNotificationSettings() returns an array and not an object

hypeJunction commented:
Yo, @Elgg/core @Elgg/contributors I have rebased this. Could I PLEASE get some eyes on this? This has a lot of database mocking logic that I can't live without.

juho-jaakkola commented:
I'll try to take a look in the weekend.

mrclay commented on engine/classes/Elgg/Notifications/NotificationEvent.php:
```diff
@@ -0,0 +1,64 @@
+	 *
+	 * @throws InvalidArgumentException
+	 */
+	public function __construct(ElggData $object = null, $action = null, ElggEntity $actor = null);
```
Interfaces usually don't specify constructors. Is this necessary?

hypeJunction replied:
No.

juho-jaakkola commented:
Nope, I ran out of weekend.

hypeJunction commented:
It's too  big of a change for me to make an executive decision, so it will have to sit here until it becomes stale. I doubt I will be revisiting this.

hypeJunction commented:
If anyone wants to finish this, please do. I am moving on.

juho-jaakkola commented:
It seems I'm 3 days too late. How exactly does #10076 affect this? Is is actively rendering the current tests obsolete, or is it just something that would now allow the tests to be written better?

hypeJunction commented:
Need to pull out the mocking logic and change the tests a bit. The
notification api will stay as is

On Aug 20, 2016 10:53 AM, "Juho Jaakkola" <notifications@github.com> wrote:

> It seems I'm 3 days too late. How exactly does #10076
> <https://github.com/Elgg/Elgg/pull/10076> affect this? Is is actively
> rendering the current tests obsolete, or is it just something that would
> now allow the tests to be written better?
>
> —
> You are receiving this because you modified the open/close state.
> Reply to this email directly, view it on GitHub
> <https://github.com/Elgg/Elgg/pull/9946#issuecomment-241188147>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/ABJaSaceMIZ8AYdnpoGxlauQUIK1UfOWks5qhsCmgaJpZM4JH6-9>
> .
>
